noodles-util/src/variant/writer/builder.rs

```diff
@@ -0,0 +1,186 @@
+use std::{
+    fs::File,
+    io::{self, Write},
+    path::Path,
+};
+
+use noodles_bcf as bcf;
+use noodles_bgzf as bgzf;
+use noodles_vcf as vcf;
+
+use crate::variant::{Compression, Format};
+
+use super::Writer;
+
+/// A variant writer builder.
+#[derive(Default)]
+pub struct Builder {
+    // None means infer on build; Some(None) means no compression explicitly set.
+    compression: Option<Option<Compression>>,
+    format: Option<Format>,
+}
+
+impl Builder {
+    /// Sets the compression of the output.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use noodles_util::variant::{self, Compression};
+    /// let builder = variant::writer::Builder::default().set_compression(Some(Compression::Bgzf));
+    /// ```
+    pub fn set_compression(mut self, compression: Option<Compression>) -> Self {
+        self.compression = Some(compression);
+        self
+    }
+
+    /// Sets the format of the output.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use noodles_util::variant::{self, Format};
+    /// let builder = variant::writer::Builder::default().set_format(Format::Vcf);
+    /// ```
+    pub fn set_format(mut self, format: Format) -> Self {
+        self.format = Some(format);
+        self
+    }
+
+    /// Builds a variant writer from a path.
+    ///
+    /// If the format or compression is not set, it is detected from the path extension.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use std::io;
+    /// use noodles_util::variant::{self, Compression, Format};
+    ///
+    /// let writer = variant::writer::Builder::default()
+    ///     .set_format(Format::Vcf)
+    ///     .set_compression(Some(Compression::Bgzf))
+    ///     .build_from_path("out.vcf.gz")?;
+    /// # Ok::<_, io::Error>(())
+    /// ```
+    pub fn build_from_path<P>(mut self, src: P) -> io::Result<Writer>
+    where
+        P: AsRef<Path>,
+    {
+        let src = src.as_ref();
+
+        let compression = self
+            .compression
+            .get_or_insert_with(|| detect_compression_from_path_extension(src));
+
+        if self.format.is_none() {
+            self.format = detect_format_from_path_extension(src, *compression);
+        }
+
+        let file = File::create(src)?;
```

Comment:
My instinct would have been to wrap this in a `io::BufWriter`, but I didn't to mirror the `alignment::Writer`.